src/rgw/rgw_lc.cc

```diff
@@ -0,0 +1,546 @@
+            }
+          }
+        }
+      } while (is_truncated);
+    }     
+  }
+
+  return ret;
+}
+
+int RGWLC::bucket_lc_post(int index, int max_lock_sec, cls_rgw_lc_obj_head& head, 
+                                                              pair<string, int >& entry, int& result)
+{
+  rados::cls::lock::Lock l(lc_index_lock_name);
+  do {
+    int ret = l.lock_exclusive(&store->lc_pool_ctx, obj_names[index]);
+    if (ret == -EBUSY) { /* already locked by another lc processor */
+      dout(0) << "RGWLC::bucket_lc_post() failed to acquire lock on, sleep 5, try again" << obj_names[index] << dendl;
+      sleep(10);
+      continue;
+    }
+    if (ret < 0)
+      return 0;
+    dout(20) << "RGWLC::bucket_lc_post()  get lock" << obj_names[index] << dendl;
+    if (result ==  -ENOENT) {
+      ret = cls_rgw_lc_rm_entry(store->lc_pool_ctx, obj_names[index],  entry);
+      if (ret < 0) {
+        dout(0) << "RGWLC::bucket_lc_post() failed to remove entry " << obj_names[index] << dendl;
+        goto clean;
+      }
+    } else if (result < 0) {
+      entry.second = lc_failed;      
+    } else {
+      entry.second = lc_complete;   
+    }
+    
+    ret = cls_rgw_lc_set_entry(store->lc_pool_ctx, obj_names[index],  entry);
+    if (ret < 0) {
+      dout(0) << "RGWLC::process() failed to set entry " << obj_names[index] << dendl;
+    }
+clean:    
+    l.unlock(&store->lc_pool_ctx, obj_names[index]);
+    dout(20) << "RGWLC::bucket_lc_post()  unlock" << obj_names[index] << dendl;
+    return 0;
+  } while (true);
+}
+
+int RGWLC::list_lc_progress(const string& marker, uint32_t max_entries, map<string, int> *progress_map)
+{
+  int index = 0;
+  progress_map->clear();
+  for(; index <max_objs; index++) {
+    map<string, int > entries;
+    int ret = cls_rgw_lc_list(store->lc_pool_ctx, obj_names[index], marker, max_entries, entries);
+    if (ret < 0)
+      return ret;
+    map<string, int>::iterator iter;
+    for (iter = entries.begin(); iter != entries.end(); ++iter) {
+      progress_map->insert(*iter);
+    }
+  }
+  return 0;
+}
+
+int RGWLC::process()
+{
+  int max_secs = cct->_conf->rgw_lc_lock_max_time;
+
+  unsigned start;
+  int ret = get_random_bytes((char *)&start, sizeof(start));
+  if (ret < 0)
+    return ret;
+
+  for (int i = 0; i < max_objs; i++) {
+    int index = (i + start) % max_objs;
+    ret = process(index, max_secs);
+    if (ret < 0)
+      return ret;
+  }
+
+  return 0;
+}
+
+int RGWLC::process(int index, int max_lock_secs)
+{
+  rados::cls::lock::Lock l(lc_index_lock_name);
+  do {
+    utime_t now = ceph_clock_now(g_ceph_context);
+    pair<string, int > entry;//string = bucket_name:bucket_id ,int = LC_BUCKET_STATUS
+    if (max_lock_secs <= 0)
+      return -EAGAIN;
+      
+    utime_t time(max_lock_secs, 0);
+    l.set_duration(time);
+
+    int ret = l.lock_exclusive(&store->lc_pool_ctx, obj_names[index]);
```

Comment:
The do while() loop can just be around the lock operation, since no other point can continue the loop.  This would make the code easier to read.